internal/pkg/config/config.go

```diff
@@ -120,31 +121,95 @@ func (c Config) IsSourceField(path, typeName, fieldName string) bool {
+	PackageRE *regexp.Regexp
+	TypeRE    *regexp.Regexp
+	FieldRE   *regexp.Regexp
+}
+
+// this type uses the default unmarshaler
+type rawSourceMatcher sourceMatcher
+
+func (s *sourceMatcher) UnmarshalJSON(bytes []byte) error {
+	raw := rawSourceMatcher{}
+	if err := json.Unmarshal(bytes, &raw); err != nil {
+		return err
+	}
+
+	// validation: do not double-specify any attribute with literal and regexp
+	if raw.Package != nil && raw.PackageRE != nil {
+		return fmt.Errorf("expected only one of Package, PackageRE to be configured")
+	}
+	if raw.Type != nil && raw.TypeRE != nil {
+		return fmt.Errorf("expected only one of Type, TypeRE to be configured")
+	}
+	if raw.Field != nil && raw.FieldRE != nil {
+		return fmt.Errorf("expected only one of Field, FieldRE to be configured")
+	}
+
+	// Copy all fields from raw
+	*s = sourceMatcher(raw)
+	return nil
 }
 
 func (s sourceMatcher) MatchType(path, typeName string) bool {
-	return s.PackageRE.MatchString(path) && s.TypeRE.MatchString(typeName)
+	return matchEither(s.Package, s.PackageRE, path) && matchEither(s.Type, s.TypeRE, typeName)
 }
 
 func (s sourceMatcher) MatchField(path, typeName, fieldName string) bool {
-	return s.PackageRE.MatchString(path) && s.TypeRE.MatchString(typeName) && s.FieldRE.MatchString(fieldName)
+	return s.MatchType(path, typeName) && matchEither(s.Field, s.FieldRE, fieldName)
 }
 
 type funcMatcher struct {
-	PackageRE  regexp.Regexp
-	ReceiverRE regexp.Regexp
-	MethodRE   regexp.Regexp
+	Package    *string
+	Receiver   *string
+	Method     *string
+	PackageRE  *regexp.Regexp
+	ReceiverRE *regexp.Regexp
+	MethodRE   *regexp.Regexp
+}
+
+type rawFuncMatcher funcMatcher
+
+func (fm *funcMatcher) UnmarshalJSON(bytes []byte) error {
+	raw := rawFuncMatcher{}
+	if err := json.Unmarshal(bytes, &raw); err != nil {
+		return err
+	}
+
+	// validation: do not double-specify any attribute with literal and regexp
+	if raw.Package != nil && raw.PackageRE != nil {
+		return fmt.Errorf("expected at most one of Package, PackageRE to be configured")
+	}
+	if raw.Receiver != nil && raw.ReceiverRE != nil {
+		return fmt.Errorf("expected at most one of Receiver, ReceiverRE to be configured")
+	}
+	if raw.Method != nil && raw.MethodRE != nil {
+		return fmt.Errorf("expected at most one of Method, MethodRE to be configured")
+	}
+
+	// copy all fields from raw
+	*fm = funcMatcher(raw)
+	return nil
 }
 
 func (fm funcMatcher) MatchFunction(path, receiver, name string) bool {
-	return fm.PackageRE.MatchString(path) && fm.ReceiverRE.MatchString(receiver) && fm.MethodRE.MatchString(name)
+	return matchEither(fm.Package, fm.PackageRE, path) &&
+		matchEither(fm.Receiver, fm.ReceiverRE, receiver) &&
+		matchEither(fm.Method, fm.MethodRE, name)
+}
+
+// TODO This is a terrible name.  matchAnyOrNil is not better.
+// Matches match against a string literal or regexp.
+// Returns vacuous true when both matchers are nil.
+func matchEither(literal *string, r *regexp.Regexp, match string) bool {
```

Comment:
I'm bad at naming, and then once I give a bad name, it's hard for me to shake it out of my head.

The `nil` handling is to be consistent with existing behavior.  If we want to say a user must specify at least one of the literal or the regexp for a given matcher attribute, we can, but that had wide-reaching impact on our own test configs.  That can be part of a future PR if you think that requirement would be good.

Reply:
Naming is hard.

Why not implement the matching for strings in terms of the matching for regexps? If you anchor the string on both ends you get a string matcher.

Reply:
From an side conversation, removed this after moving to a polymorphic underlying implementation.